simulator.go

```diff
@@ -202,13 +204,46 @@ type simulatorAPIHandler struct {
 	overrides []string
 	autoID    uint32
 
-	runner *docker.Container
-	nodes  map[string]*docker.Container
+	runner       *docker.Container
+	nodes        map[string]*docker.Container
+	nodesTimeout map[string]time.Time
 
 	result *simulationResult
 	lock   sync.RWMutex
 }
 
+// CheckTimeout is a goroutine that checks if the timeout has passed and stops
+// container if it has.
+func (h *simulatorAPIHandler) CheckTimeout() {
+	for {
+		for id, c := range h.nodes {
+			h.lock.Lock()
```

Comment:
I think this is racy -- you should probably `RLock` h.nodes. Might be less hassle to simply lock it outside the loop.

Reply:
I moved the locks outside the loop. I'm not sure if they will take up too much time when it loops through all the containers to check for the timeout. If it does, we can likely do something with RLock and such, but yeah, it's complicated. I usually try to avoid mutex's for the most part (the share memory by communicating thing). Just let me know if you want me to change something.